android/src/main/java/io/ably/lib/rest/Push.java

```diff
@@ -0,0 +1,247 @@
+        // activateWithToken with it.
+        activateWithToken(context, regToken);
+    }
+
+    public void onNewRegistrationToken(Context context, RegistrationTokenType type, String token) {
+        RegistrationToken persisted = RegistrationToken.persisted(context);
+        if (persisted != null && persisted.type != type) {
+            Log.e(TAG, "trying to register device with " + type + ", but it was already registered with " + persisted.type);
+            return;
+        }
+
+        RegistrationToken regToken = RegistrationToken.persistAndReturn(context, type, token);
+
+        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+        if (!prefs.getBoolean(SharedPrefKeys.ACTIVATING, false)) {
+            // If not ACTIVATING, ie. if activate wasn't called, just let the token in storage and
+            // be done here; when activate is called, it will pick up the token and call
+            // activateWithToken.
+            return;
+        }
+
+        activateWithToken(context, regToken);
+    }
+
+    private void activateWithToken(final Context context, RegistrationToken token) {
+        // Preconditions: ACTIVATING = true; ACTIVATED = false
+        // Postconditions: ACTIVATING = false; ACTIVATED = true iff succeeded (token renewed or new registration)
+        if (token == null) {
+            // If there's no token, onNewRegistrationToken will call this again with the token.
+            return;
+        }
+
+        final DeviceDetails device = rest.device != null ? rest.device : DeviceDetails.loadFromDevice(context, rest);
+
+        device.setRegistrationToken(token);
+
+        boolean isNewRegistration = device.updateToken == null;
+        if (isNewRegistration) {
+            Http.RequestBody body = rest.http.requestBodyFromGson(device.toJsonObject());
+            rest.asyncHttp.post("/push/deviceRegistrations", HttpUtils.defaultAcceptHeaders(rest.options.useBinaryProtocol), null, body, new Serialisation.HttpResponseHandler<JsonObject>(), new Callback<JsonObject>() {
+                @Override
+                public void onSuccess(JsonObject response) {
+                    device.setAndPersistUpdateToken(context, response.getAsJsonPrimitive("updateToken").getAsString());
+                    rest.device = device;
+                    new ActivateCallback(context, "PUSH_ACTIVATED").onSuccess(null);
+                }
+                @Override
+                public void onError(ErrorInfo reason) {
+                    new ActivateCallback(context, "PUSH_ACTIVATED").onError(reason);
+                }
+            });
+        } else {
+            rest.device = device;
+            Http.RequestBody body = rest.http.requestBodyFromGson(device.pushMetadataJsonObject());
+            rest.asyncHttp.put("/push/deviceRegistrations/" + device.id, HttpUtils.defaultAcceptHeaders(rest.options.useBinaryProtocol), null, body, null, new ActivateCallback(context, "PUSH_RENEWED_TOKEN"));
```

Comment:
Yeah, that's something I've considered here but I think it's preferable to just call the callback (the BroadcastReceivers in this case) with an error, just like we do with all other requests. What can we really do if all fallback hosts, etc. fail? It's probably the user who needs to fix something (auth, etc.).